// freepbx/dist/lib/Conf.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _bluebird = require('bluebird');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Conf = function () {
	function Conf(db) {
		_classCallCheck(this, Conf);

		this.configs = {};
		this.db = db;
		this.init = false;
	}

	_createClass(Conf, [{
		key: 'load',
		value: function load() {
			var _this = this;

			return new _bluebird.Promise(function (resolve, reject) {
				var query = _this.db.query('SELECT * FROM freepbx_settings');
				query.on('result', function (row) {
					var val = row.value;
					if (row.type === "bool") {
						val = row.value !== 0 ? true : false;
					}
					_this.configs[row.keyword] = val;
				}).on('end', function () {
					_this.init = true;
					resolve(_this.configs);
				});
			});
		}
	}, {
		key: 'getAll',
		value: function getAll() {
			var _this2 = this;

			var cached = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;

			if (!this.init || !cached) {
				return new _bluebird.Promise(function (resolve, reject) {
					return _this2.load().then(function (configs) {
						resolve(configs);
					});
				});
			} else {
				return _bluebird.Promise.resolve(this.configs);
			}
		}
	}, {
		key: 'get',
		value: function get(keyword) {
			var _this3 = this;

			var cached = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;

			if (!cached) {
				return new _bluebird.Promise(function (resolve, reject) {
					_this3.load().then(function (configs) {
						if (typeof _this3.configs[keyword] !== "undefined") {
							resolve(_this3.configs[keyword]);
						} else {
							reject(new Error('configs is undefined'));
						}
					});
				});
			} else {
				if (typeof this.configs[keyword] !== "undefined") {
					return _bluebird.Promise.resolve(this.configs[keyword]);
				} else {
					return _bluebird.Promise.reject(new Error('configs is undefined'));
				}
			}
		}
	}]);

	return Conf;
}();

exports.default = Conf;
